File: musicpimp/src/main/scala/org/musicpimp/ui/receivers/RemoteControlReceiver.scala
package org.musicpimp.ui.receivers

import android.content.{Intent, Context, BroadcastReceiver}
import android.view.KeyEvent
import org.musicpimp.audio.PlayerManager

/** Handles hardware key presses of prev/next/pause/play, if any.
  */
class RemoteControlReceiver extends BroadcastReceiver {
  def player = PlayerManager.active

  def onReceive(context: Context, intent: Intent) {
    intent.getAction match {
      case Intent.ACTION_MEDIA_BUTTON =>
        val event = intent.getParcelableExtra(Intent.EXTRA_KEY_EVENT).asInstanceOf[KeyEvent]
        event.getKeyCode match {
          case KeyEvent.KEYCODE_MEDIA_PLAY =>
            player.resume()
          case KeyEvent.KEYCODE_MEDIA_PAUSE =>
            player.pause()
          case KeyEvent.KEYCODE_MEDIA_NEXT =>
            player.playNext()
          case KeyEvent.KEYCODE_MEDIA_PREVIOUS =>
            player.playPrevious()
        }
    }
  }
}

File: app/src/main/java/org/musicpimp/media/RemoteControlReceiver.kt
package org.musicpimp.media

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.view.KeyEvent
import org.musicpimp.PimpComponents
import org.musicpimp.audio.Player

/** Handles playback commands from external devices.
 */
class RemoteControlReceiver(private val comps: PimpComponents) : BroadcastReceiver() {
    companion object {
        private val intentFilter = IntentFilter(Intent.ACTION_MEDIA_BUTTON)

        fun register(comps: PimpComponents) =
            comps.appContext.registerReceiver(RemoteControlReceiver(comps), intentFilter)
    }

    val player: Player get() = comps.player

    override fun onReceive(context: Context, intent: Intent) {
        when (intent.action) {
            Intent.ACTION_MEDIA_BUTTON -> {
                val event = intent.getParcelableExtra(Intent.EXTRA_KEY_EVENT) as KeyEvent
                when (event.keyCode) {
                    KeyEvent.KEYCODE_MEDIA_PLAY ->
                        player.resume()
                    KeyEvent.KEYCODE_MEDIA_PAUSE ->
                        player.pause()
                    KeyEvent.KEYCODE_MEDIA_NEXT ->
                        player.next()
                    KeyEvent.KEYCODE_MEDIA_PREVIOUS ->
                        player.prev()
                }
            }

        }
    }
}
